pkg/services/ngalert/state/manager.go

```diff
@@ -304,6 +312,10 @@ func (st *Manager) Put(states []*State) {
 
 // TODO: Is the `State` type necessary? Should it embed the instance?
 func (st *Manager) saveAlertStates(ctx context.Context, states ...*State) (saved, failed int) {
+	if st.instanceStore == nil {
```

Comment:
nit: I'm a big fan of the [null object pattern](https://en.wikipedia.org/wiki/Null_object_pattern), you can get rid of all these conditionals if you introduce it.

Reply:
I also like this approach. I used this approach to prevent the manager from emitting logs related to stubbed operations. We can refactor this further